### DISCUSSION TITLE
Improve or replace Tinytest

### DISCUSSION THREAD

**Discussion by jankapunkt** [meteor/meteor#12162] [Feature Requests]
I know this test package is lightweight and spread all over but there are two things that make it very hard to work with:

- headless client testing (like with mocha + puppeteer) is not easily done without further additional code
- there is no code coverage so I can't know for sure how much a package is covered by the tests
- this is the main reason why I try to avoid to use it for packages and nearly always use `meteortesting:mocha`

The overall stability would profit from this imo

**Comment by zodern** [comment]
> headless client testing (like with mocha + puppeteer) is not easily done without further additional code

I created an npm package to make it easy to run tinytest headlessly: https://www.npmjs.com/package/@zodern/mtest/v/0.5.1.

I mostly like tinytest, and still use it for my packages.  The main issue I have with it is sometimes if there is an error on the client, it will only run server tests and show everything as passing even though client tests never ran.

**Reply by jankapunkt** [reply]
Hey @zodern that's great and I didn't know this exists! With this we can now run packages tests using github ci. Do you think it's possible to integrate coverage report via mtest, too?

**Reply by jankapunkt** [reply]
@zodern would that also allow to introduce coverage reports for `Tinytest`?

**Reply by zodern** [reply]
I've never looked into it. It uses `meteor test-packages` to run the tests, so if there is a way to get coverage with that then it would work with mtest.

**Comment by StorytellerCZ** [comment]
We could potentially replace Tinytests with Node Tests: https://nodejs.org/docs/latest-v20.x/api/test.html
Preferably once we have Node 20 in Meteor.

**Reply by italojs** [reply]
It will be a dream coming true for me lol

**Reply by dupontbertrand** [reply]
Maybe I can work on that and show something in a few days/weeks :)

**Reply by italojs** [reply]
I have a [poc](https://github.com/italojs/meteor-node-test-runner/tree/main/meteor-node-test-runner) where I enable apps to use node-native test-runner but cant make it work at meteor source

**Comment by italojs** [comment]
I don't know much about the tiny test history. I assume Mocha or Jest wasn't good enough in ~2015 for Meteor, but today it's clear it's like fibers in Meteor 2.x; it's a thorn in our shoes. 

BTW, why do we still uses tinytest? aside from the amount of work, do we have any limitations that prevent us from moving to the Native Node test runner or other popular libraries like Jest or Mocha?

**Reply by jankapunkt** [reply]
Please also consider test coverage and headless testing being so much easier using mocha

**Reply by italojs** [reply]
for headless tests I pretend to use puppeteer + native node test runner to assert/mock the code, do you know any browser headless feature supported by mocha?

**Reply by italojs** [reply]
I was reflecting on this weekend, and I think it's better to keep the "tiny test" for a while and just improve it (we need parallel tests asap) since it's a pretty stable test runner. Replacing it within the Meteor test environment would be a challenge in terms of effort, and we can spend that effort on something else (use npm instead of Atmosphere, use node from environment instead of Meteor's node bin, those things).

**Comment by ceigey** [comment]
I wish I understood how test drivers worked better. Since I've been reliant on Meteor Vite for now, I've had this pipe-dream of somehow finding an "easier than expected" way of using Vitest with Meteor, but I don't know enough really to make it work.

Rstack has their own proposed Vitest equivalent, Rstest: https://rstest.rs/guide/start/quick-start

> Rstest now supports core testing scenarios including Node / DOM testing, mocking, multi-project testing, and coverage collection, and has been successfully implemented in Rstack projects (including Rsbuild, Rslib, Rspack, and others

Can see their roadmap here: https://github.com/web-infra-dev/rstest/issues/85

^ could be something solid to build a TinyTest v2 on. E.g. expose the same API to make migration seamless, but underneath it's really just something more commonplace so the core team can just focus on integrating with Meteor's runtime (same could apply to any test runner though).

**Comment by italojs** [comment]
I tried to replace it (but gave up when I realized how much work it would be) and created a package that allows us to use the Node native test runner for Meteor apps. It's at version 0.0.3 but already executes tests for server, client, and in parallel. I based it on [meteor-mocha](https://github.com/Meteor-Community-Packages/meteor-mocha).

## [italojsmeteor:node-test-runner](https://github.com/italojs/meteor-node-test-runner/tree/main/meteor-node-test-runner)

Meteor driver that runs server and client suites with the **Node.js native test runner** (no fibers), reusing the headless plumbing from `meteortesting:browser-tests`.

it supports parallel tests and (server || client) only

### Installation and Usage

Install the package in your Meteor app:
```bash
meteor add italojsmeteor:node-test-runner
```

Then run your tests:
```bash
# Run all tests (server + client)
TEST_BROWSER_DRIVER=puppeteer meteor test --driver-package italojsmeteor:node-test-runner
```

Please, test it and share your feedback, it can be a start point to replace the tiny test one day from Meteor's source.

in the repository root we have the package + a test demo where you can run test the pkg

**Comment by dupontbertrand** [comment]
## A staged plan to put `node:test` under Meteor — and a foundation that already works

The wish to improve or replace Tinytest keeps surfacing — here, in issues, and in forum
threads. The recurring themes are always the same: **no code coverage**, **headless /
client testing is painful**, **assertions and tooling feel dated**, and newcomers have
to learn a Meteor-only API. Tinytest is stable and lightweight, but its DX is a real
drag on the ecosystem's test culture.

Rather than open *yet another* "we should…" comment, I'd like to actually drive this as
a **staged effort**, and I've already built the first stage so we can talk about
something concrete. The guiding principle is **zero disruption at every step**:
opt-in, additive, nothing removed until the very end.

### The plan

```mermaid
flowchart LR
    S0["**Stage 0**<br/>test-in-node driver<br/>opt-in · server-side<br/>(built — see PR)"]
    B["**Bridge**<br/>run existing Tinytest<br/>tests unchanged"]
    M["**Pilot migration**<br/>one core package<br/>as a worked example"]
    BR["**Browser path**<br/>real headless story<br/>for DOM tests"]
    D["**Deprecation**<br/>retire Tinytest<br/>once adoption + browser land"]
    S0 --> B --> M --> BR --> D
    classDef done fill:#dcffe4,stroke:#2da44e;
    class S0 done;
```

Each box is a **separate, independently reviewable PR**. We never have a "big bang":
Tinytest stays fully functional the whole way, and is only deprecated at the end, after
the gaps (notably browser testing) are genuinely covered.

### Why `node:test`, and why it composes with Meteor

The key technical reason it's viable: **isobuild owns module resolution** (`meteor/*`
are not real npm modules on disk). That makes the choice of runner not a matter of taste
but of architecture:

```mermaid
flowchart TB
    ISO["isobuild owns module resolution<br/>(meteor/* resolved by the bundler)"]
    subgraph P["Passive runners — compose ✅"]
      NT["node:test"]
      MO["Mocha"]
    end
    subgraph A["Active runners — conflict ❌"]
      JE["Jest"]
      VI["Vitest / Rstest"]
    end
    P -->|"run inside the already-compiled process"| ISO
    A -.->|"want to own transform + resolution + mock hoisting"| ISO
```

Passive runners (node:test, Mocha) execute tests inside the process isobuild already
compiled — they don't fight for control. Active runners (Jest, Vitest, Rstest) want to
own transforms, module resolution and mock hoisting, which collides head-on with
isobuild. That's why a Mocha driver exists in the community but a maintained Jest driver
doesn't. `node:test` has the additional advantage of being **built into Node** (Node 22):
coverage, mocking, snapshots, filtering, sharding and multiple reporters, **zero
dependencies**, and isobuild passes `node:` imports through untouched on the server arch
— so no build-system changes are needed.

### The browser question, addressed head-on

The single hardest part — and the one this plan does **not** hand-wave — is client /
browser testing. An audit of the core packages that have "client" tests shows the split:

```mermaid
flowchart TD
    T["Package tests labelled 'client'"] --> Q{"Actually touches the DOM?"}
    Q -->|"No — the large majority:<br/>LocalCollection, query operators,<br/>EJSON, Tracker, check, session…"| N["Pure logic →<br/>node:test, server-side<br/>(Stage 0, available now)"]
    Q -->|"Yes — a small minority:<br/>real DOM, localStorage,<br/>runtime-client checks"| BR["Genuinely needs a browser →<br/>headless browser path<br/>(planned step, not pre-empted)"]
```

Most "client" tests are pure JavaScript logic that never touches `document` or `window`
— those can already run server-side under `node:test` today. The minority that truly
needs a browser keeps a **real** headless path (Puppeteer-style), planned as its own
step. `node:test` runs the logic; the browser path covers the DOM. Tinytest's browser
testing stays intact until that path is in place.

### Stage 0 — already built and opt-in

The foundation is a small core driver, `test-in-node`, opt-in and server-side:

```bash
meteor test-packages my-package --driver-package test-in-node --once
```

Tests are plain Node tests:

```js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { Random } from 'meteor/random';

describe('my-package', () => {
  it('generates unique ids', () => {
    assert.notStrictEqual(Random.id(), Random.id());
  });
});
```

The only interesting bit is **completion detection**, because Meteor keeps the event loop
alive (HTTP + Mongo) so Node's `beforeExit` never fires — which means `node:test` never
emits its final summary. The driver solves this with a **structured** mechanism (a
`node:test` reporter for counts + a root `after()` hook as the completion barrier; it
finalizes once every enqueued test has completed). No stdout scraping, no debounce, no
`beforeExit` dependency. Details and the full diff are in the PR: **#14477**.

### How this answers the concerns already raised here

- **Code coverage** → native, `--experimental-test-coverage` (the original ask in this
  thread).
- **"Replacing Tinytest is too much effort"** → the **bridge** (step 1) runs *existing*
  Tinytest tests unchanged by re-registering them as `node:test` cases. Adoption is
  incremental and costs ~nothing per package.
- **"Keep the same API so migration is seamless"** → that's exactly what the bridge does.
- **Headless / client testing** → kept, via the browser-path step; not pre-empted, and
  the bulk of "client" tests already run server-side (diagram above).
- **Jest / Vitest / Rstest** → great tools, but active runners that conflict with
  isobuild (diagram above); `node:test` is the passive runner that composes.

There's also good prior art in the community for headless Tinytest and Node-runner
drivers — this plan aims to land the same capability **in core**, on the real `node:test`,
in reviewable steps.

### What I'm asking

I've explored this fairly deeply (earlier PRs #14216 and #14230 were research toward it);
this plan extracts the clean, minimal foundation and sequences the rest so each piece is
discussable on its own. **Stage 0 is open as a PR: #14477.** I'd love feedback on:

1. the staged direction overall (is the sequencing right?),
2. the Stage 0 driver as the foundation,
3. the browser-path approach for the ~minority of DOM tests.

Happy to drive this end-to-end cc @nachocodoner @italojs

**Reply by italojs** [reply]
The tinytest bridge isn't clear for me, will it be a node:test wrapper following the tinytest APIs? it could be a good idea, but we should care about the limitations. For example, Tinytest can run tests in parallel today, so the bridge shoulnd have this kind of limitation

**Reply by dupontbertrand** [reply]
Yes — the bridge is exactly that: a thin `node:test` wrapper over the Tinytest API. It reads the registered cases (`Tinytest._TestManager.ordered_tests`) and re-registers each as a `node:test` `it()`, handing each test a `test` object whose methods (`test.equal`, `test.isTrue`, `test.throws`, …) delegate to `node:assert`. Existing package tests run unchanged; `node:test` is just the engine underneath. (Prototyped in the now-closed #14230.)

On parallelism — good thing to pin down. Tinytest's package-test runner actually executes **sequentially** today: the `runNext` loop in `tinytest.js` runs one case, then the next on completion — no `Promise.all`/concurrency. `node:test` has built-in concurrency (the `concurrency` option on suites/tests), so this is a capability we'd **gain**, not lose.

For the bridge specifically, the one real constraint is isolation: a lot of existing Tinytest tests assume sequential execution and shared DB state (the runner clears collections between/around them). So I'd keep bridged *legacy* tests sequential-safe by default, with concurrency opt-in per suite, while *natively written* `node:test` tests get parallelism straight away. Net: legacy correctness preserved, parallelism unlocked.

The bridge is its own follow-up PR (step 1 of the plan) — happy to design the parallel-execution story there together, since you've already been down this road with `node-test-runner`. The current PR (#14477) is just the Stage 0 driver foundation.

**Reply by italojs** [reply]
ok, it brings me 2 questions:

1. Why not just use a TinyTest wrapper for everything instead use 2 test approaches? 

2. It smells like a workaround that will be there forever (IMO). Did you evaluate any other option we can use for both scenarios (server and client)? If not, we can create a lib for that together inside Meteor Org

**Reply by dupontbertrand** [reply]
These two are related, so together:

**1. Why not just a TinyTest wrapper for everything?**

It's not really 2 approaches — it's **one target** (native `node:test` + `node:assert`) and **one temporary off-ramp** (the bridge), which converge to a single surface and ends with the bridge being deprecated. The real question is which surface is permanent. I'd argue against "TinyTest wrapper forever" because:

- it doesn't solve what this thread is about — the complaints (Meteor-only API, dated assertions, no standard tooling) are about the *surface*, not the engine; wrapping in TinyTest keeps `test.equal` instead of `assert.strictEqual`, knowledge that transfers nowhere;
- the TinyTest API caps us at its ceiling: no `describe`/nesting, no `before`/`after`, no `t.mock`, no snapshots, no name filtering — native gives all of that for free;
- "wrapper for everything" is actually the *more* permanent workaround (it bakes in `_TestManager` forever); the bridge is the honest, time-boxed version with a deprecation path.

**2. Any option that works for both server and client?**

Yes — `node:test` *is* that single option, because it's a **passive** runner: server vs client is an environment split, not a runner split. The same test code runs in Node for the server arch and in a real headless browser for the client arch. One engine, two environments.

For the client side I evaluated two paths: DOM emulation in Node (happy-dom — built it in the closed #14230; fast, but not a real browser, so dishonest for true-DOM tests), and a **real headless browser** via the existing `meteortesting:browser-tests` plumbing — which is exactly what your `node-test-runner` already does. We've independently landed on the same substrate.

So **yes, let's build this in the Meteor Org together.** The "browser path" in my plan *is* that lib; Stage 0 (#14477) is just its server-side foundation, not a competing third thing.

**Reply by italojs** [reply]
Hmm, having two paths for these tests still bothers me. I really believe that temporary implementations can easily become permanent.

Let me stress and expand on this a bit. Let's start with our core necessities and current pain points:

* **Isomorphic Promise:** Meteor must guarantee that tests run on both the server and the browser (multi-browser if possible).
* **Parallelism Blocker:** Meteor has a massive test suite, but `tinytest` doesn't allow us to run tests in parallel.
* **Environment Isolation:** Currently, the Meteor test server is shared across most tests, which complicates environment isolation (and directly hinders parallelism).
* **Migration Cost:** With so many existing tests, rewriting all of them at once is an unrealistic challenge.
* **Over-engineered Wrapper:** `tinytest` does too many things—it isn't just a test wrapper, but a full test framework embedded inside another framework (Meteor).

Because of the `isobuild` limitation, tools like `rspack` or `jest` are out of the game. At the same time, implementing just `node:test` won't solve our parallelism issues due to the single-server setup, and it won't let us test on Bun or Deno in the future (I hope we can support multiple runtimes someday).

### A Definitive Path (Debatable)

I see a path by creating a new test layer (or rewriting the current one) that strictly follows the `tinytest` API. This approach would allow us to:

* **Support modern runners:** Implement `node:test` and other runtime test-runners (maybe use [poku](https://poku.io/) cc: @wellwelwel).
* **Ensure test isolation:** Set up one Meteor server per test (or per worker, since spawning a Meteor process is expensive), or ensure a strict cleanup of the test server before each execution.
* **Unlock parallelism:** Run all of them in parallel.
* **Expand coverage:** Run all of them in different browsers.
* **Backward compatibility:** Drastically minimize the number of required changes to our current tests.

What do you think? It follows your suggestion but in a more definitive way. 

But before implementing it, let's discuss the details—I'm afraid to waste your code-time lol.

**Reply by dupontbertrand** [reply]
I think we are mostly aligned here.

One correction on my side: earlier I implied that native `node:test` tests would give us parallelism almost immediately. That is not really true. `node:test` gives us the concurrency primitive, but Meteor’s current shared server + Mongo test environment is still the real bottleneck. So I agree: the API alone does not solve parallelism. The isolation model does.

Your point about multi-runtime also makes sense to me. Writing the whole new path directly against `node:test` would make Node the only long-term target. A better design is probably a small runner seam, with `node:test` as the first/reference adapter, but not the only possible one.

That also matches the kind of boundary I have been exploring in a few Meteor PRs, with different maturity levels:

* pluggable WebSocket transport — #14231, merged;
* pluggable DDP serializer — #14235, experimental / discussion material;
* runtime / ESM work around Node and Bun — #14312/#14313, still draft / exploratory.

So I do not want to overstate this as an already-settled Meteor direction. But the pattern seems useful: keep the Meteor-specific boundary stable, and allow the underlying implementation to vary behind a thin adapter.

For tests, that could mean:

* a thin runner adapter layer: `node:test`, maybe `poku`, maybe `bun:test` / Deno later;
* a Tinytest compatibility adapter, so the existing test suite does not need a big-bang rewrite;
* a separate orchestration layer above that for Meteor-specific concerns.

I think that separation matters. The runner seam should answer “which test engine executes this test?”. It should not try to become a new testing framework.

The harder Meteor-specific parts live above it:

* server/process isolation;
* one Mongo database per worker;
* parallelism at worker level rather than per test;
* browser fan-out for real client tests;
* aggregation/reporting back to the Meteor tool.

So I would see `node:test` as the first practical implementation, not the whole architecture.

In that framing, #14477 is still useful, but only as Stage 0: an opt-in, server-side, single-process foundation. No claim that it solves parallelism yet. It just gives us a minimal place to start wiring this properly.

If we agree on that split — thin runner seam + Tinytest compatibility + Meteor orchestration above — I can write a small RFC around it before pushing the implementation further. Then we can discuss the interface shape and the isolation model explicitly.

**Reply by italojs** [reply]
Yes, please, go ahead with that RFC, I like it, but try to emphasize the parallelism; all this work front will come from this pain, so the parallelism should be present in the early PRs, it will reduce our CI time

**Reply by jankapunkt** [reply]
One thing that remains an open question to me when using `node:test` is, how coverage will be able to backtrack from the isobuild-compiled code to the original sources? Source maps? From my understanding the "not relying on isobuild" only applies to the node test runner itself, right?

**Reply by italojs** [reply]
just attaching it here https://rstest.rs/guide/advanced/playwright
note: I still didn't reflect about it

**Comment by dupontbertrand** [comment]
**The RFC is up: "A parallel-first Meteor test runner platform"** — full text (with revision history): https://gist.github.com/dupontbertrand/51cefeafb4aab3de9f191681ecdbf2c5

@italojs — as requested when you greenlit it: **parallelism is the headline, not a late-stage bonus**. Stage 1 (the first PR after #14477) is the worker orchestration that cuts CI time. And the pain is live — #14606 had to add fail-fast handling for stalled `test-in-console` suites just last week.

**TL;DR** (from the RFC):

- The pain is **CI time**: the test suite can't run in parallel.
- The real blocker is **not Tinytest's API alone** — it's the **single shared server process + single shared Mongo database** every run uses.
- The fix is a thin **runner seam** (`node:test` first) plus a Meteor **orchestration layer** that isolates **one process + one Mongo DB per worker**.
- **Stage 1 delivers the parallelism (and the CI-time win) early.** Existing Tinytest tests keep working via a **compatibility adapter**; nothing is removed until the browser gap is closed.
- This is the shape the ecosystem converged on in 2025-2026 — **the framework owns harness + orchestration, a commodity runner executes** (Rails, Angular v21, Nuxt) — and Playwright 1.62's newly stable component testing gives the browser path a passive integration too.

Three things worth flagging since our last exchange:

1. **@italojs — your Rstest link is addressed, and it fits.** Rstest is named in §4.1 as a credible **second engine** behind the seam once two gates clear: Rstest reaching 1.0 (it's 0.11.x with weekly breaking changes as of July), and Meteor's Rspack config becoming the authority on app compilation (post-3.6). Its Playwright-based browser mode is likewise a candidate engine for the Stage 4 browser path. Notably, Rstest's own default pool is fork-per-file **process isolation** — the same model this RFC proposes, independently arrived at.
2. **@jankapunkt — your coverage/source-map question is answered with measurements** (§10.7): I ran a spike on #14477. Short version: V8 coverage is real but attributes to the linked `meteor://` virtual files, not original sources; isobuild already emits the needed `.map` per server file; the remap is wiring — the same standalone `ast-v8-to-istanbul` package Vitest 4 uses by default — owned by the orchestration layer, not a promise on Node's still-experimental coverage flag.
3. **The browser path got dramatically cheaper on July 24.** Playwright 1.62 stabilized component testing as a dev-server-agnostic "stories and galleries" model — *"if your dev server can render it, Playwright can test it."* Meteor's own dev server can serve that gallery from the isobuild build: real Chromium/Firefox/WebKit client tests through a **passive** integration (§7, Stage 4). This directly addresses the isomorphism concern raised on #14230.

Happy to iterate on any section — amendments will show in the gist's revision history.
